### PR TITLE
error-msg: improve error message output

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
[TOOL-1641](https://internal.pingcap.net/jira/browse/TOOL-1641)
Some error messages of DM are confusing and not clear.

### What is changed and how it works?
Refine error message and info.

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the `dm/dm-ansible`
 - Need to be included in the release note

### CODE REVIEW

File: pkg/parser/common.go [hunk 1]
```diff
@@ -32,7 +33,11 @@ import (
 func Parse(p *parser.Parser, sql, charset, collation string) (stmt []ast.StmtNode, err error) {
 	stmts, warnings, err := p.Parse(sql, charset, collation)
 	if err != nil {
-		log.L().Error("parse statement", zap.String("sql", sql), log.ShortError(err))
+		if utils.IsBuildInSkipDDL(sql) {
+			log.L().Warn("parse statement", zap.String("sql", sql), log.ShortError(err))
```
Comment (WangXiangUSTC) [inline pkg/parser/common.go]: seems useless, although this ddl should be skipped, will also return error

File: pkg/streamer/reader.go [hunk 2]
```diff
@@ -202,6 +202,14 @@ func (r *BinlogReader) parseDirAsPossible(ctx context.Context, s *LocalStreamer,
 				if !strings.HasSuffix(relayLogFile, pos.Name) {
 					return false, "", "", terror.ErrFirstRelayLogNotMatchPos.Generate(relayLogFile, pos)
 				}
+				relayFilepath := path.Join(dir, relayLogFile)
+				cmp, err := fileSizeUpdated(relayFilepath, offset)
+				if err != nil {
+					return false, "", "", err
+				}
+				if cmp < 0 {
+					return false, "", "", terror.ErrRelayLogGivenPosTooBig
```
Comment (WangXiangUSTC) [inline pkg/streamer/reader.go]: there are two `for` loops in this function, so this error may not be caused by the meta file.

File: pkg/parser/common.go (hunk 1 above)
Comment (lichunzhu) [reply pkg/parser/common.go]: Do you mean that the outer function will print this error?

Comment (WangXiangUSTC) [reply pkg/parser/common.go]: yes

File: pkg/streamer/reader.go (hunk 2 above)
Comment (lichunzhu) [reply pkg/streamer/reader.go]: Add a new function `checkRelayPos` to do that before loop.

File: pkg/parser/common.go (hunk 1 above)
Comment (lichunzhu) [reply pkg/parser/common.go]: Revised. The error log will be printed when function `CheckIsDDL` of relay module is deployed. Now we will check the skipped ddl before parse in `CheckIsDDL`.

**Comment by lichunzhu** [comment]
@WangXiangUSTC Revised, PTAL again

**Comment by WangXiangUSTC** [comment]
rest LGTM

**Comment by lichunzhu** [comment]
@WangXiangUSTC Revised, PTAL
@csuzhangxc PTAL

**Comment by WangXiangUSTC** [comment]
LGTM

File: pkg/terror/error_list.go [hunk 3]
```diff
@@ -815,7 +818,7 @@ var (
 	ErrMasterQueryWorkerConfig     = New(codeMasterQueryWorkerConfig, ClassDMMaster, ScopeInternal, LevelHigh, "")
 	ErrMasterOperNotFound          = New(codeMasterOperNotFound, ClassDMMaster, ScopeInternal, LevelHigh, "operation %d of task %s not found, please execute `query-status` to check status")
 	ErrMasterOperRespNotSuccess    = New(codeMasterOperRespNotSuccess, ClassDMMaster, ScopeInternal, LevelHigh, "operation not success: %s")
-	ErrMasterOperRequestTimeout    = New(codeMasterOperRequestTimeout, ClassDMMaster, ScopeInternal, LevelHigh, "request is timeout, but request may be successful, please execute `query-status` to check status")
+	ErrMasterOperRequestTimeout    = New(codeMasterOperRequestTimeout, ClassDMMaster, ScopeInternal, LevelHigh, "request to worker is timeout, but request may be successful, please execute `query-status` to check status")
```
Comment (csuzhangxc) [inline pkg/terror/error_list.go]: 1. should we need to replace `worker` with `dm-worker`
2. should we need to include the IP:port of the DM-worker in the error message?

File: pkg/terror/error_list.go [hunk 4]
```diff
@@ -516,7 +518,8 @@ var (
 	ErrWatcherStart             = New(codeWatcherStart, ClassFunctional, ScopeInternal, LevelHigh, "watcher starts for relay log dir %s")
 	ErrWatcherChanClosed        = New(codeWatcherChanClosed, ClassFunctional, ScopeInternal, LevelHigh, "watcher's %s chan for relay log dir %s closed")
 	ErrWatcherChanRecvError     = New(codeWatcherChanRecvError, ClassFunctional, ScopeInternal, LevelHigh, "watcher receives error, relay log dir %s")
-	ErrRelayLogFileSizeSmaller  = New(codeRelayLogFileSizeSmaller, ClassFunctional, ScopeInternal, LevelHigh, "file size of relay log %s become smaller")
+	ErrRelayLogGivenPosTooBig   = New(codeRelayLogGivenPosTooBig, ClassFunctional, ScopeInternal, LevelHigh, "the relay log pos of meta config is too big, please check it again")
```
Comment (csuzhangxc) [inline pkg/terror/error_list.go]: 1. now, we arrange `Err*` and `code*` in the same order. how about keeping the order as before?
2. can we including `relay log pos of meta config` in the error message?

File: pkg/terror/error_list.go [hunk 5]
```diff
@@ -516,7 +518,8 @@ var (
...
+	ErrRelayLogFileSizeSmaller  = New(codeRelayLogFileSizeSmaller, ClassFunctional, ScopeInternal, LevelHigh, "file size of relay log %s become smaller, please check the status of relay log and re-pull it")
```
Comment (csuzhangxc) [inline pkg/terror/error_list.go]: how should the user `re-pull` binlog?

File: syncer/syncer.go [hunk 6]
```diff
@@ -1302,7 +1302,7 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 					s.tctx.L().Info("re-replicate shard group was completed", zap.String("event", "XID"), zap.Reflect("re-shard", shardingReSync))
 					err = closeShardingResync()
 					if err != nil {
-						return err
+						return terror.Annotatef(err, "current pos %s", currentPos)
```
Comment (csuzhangxc) [inline syncer/syncer.go]: this `currentPos` has not to be updated yet (updated in L1312).

**Comment by lichunzhu** [comment]
@csuzhangxc All comments have been addressed. PTAL.

File: dm/master/server.go [hunk 7]
```diff
@@ -356,6 +356,8 @@ func (s *Server) OperateTask(ctx context.Context, req *pb.OperateTaskRequest) (*
 
 	var wg sync.WaitGroup
 	for _, worker := range workers {
+		// https://golang.org/doc/faq#closures_and_goroutines
+		worker := worker
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply dm/master/server.go]: The `worker` in L382 works. This is mainly for L337. It will take some time for go routine to use `worker`.

**Comment by lichunzhu** [comment]
Updated. PTAL @csuzhangxc

File: pkg/terror/error_list.go [hunk 8]
```diff
@@ -516,7 +518,7 @@ var (
 	ErrWatcherStart             = New(codeWatcherStart, ClassFunctional, ScopeInternal, LevelHigh, "watcher starts for relay log dir %s")
 	ErrWatcherChanClosed        = New(codeWatcherChanClosed, ClassFunctional, ScopeInternal, LevelHigh, "watcher's %s chan for relay log dir %s closed")
 	ErrWatcherChanRecvError     = New(codeWatcherChanRecvError, ClassFunctional, ScopeInternal, LevelHigh, "watcher receives error, relay log dir %s")
-	ErrRelayLogFileSizeSmaller  = New(codeRelayLogFileSizeSmaller, ClassFunctional, ScopeInternal, LevelHigh, "file size of relay log %s become smaller")
+	ErrRelayLogFileSizeSmaller  = New(codeRelayLogFileSizeSmaller, ClassFunctional, ScopeInternal, LevelHigh, "file size of relay log %s become smaller, please check the status of relay log and re-pull it. If you want to re-pull it, you should open relay.meta, set the binlog-name to the error pos name, set binlog-pos to 4, delete the stashed relay log and run `resume-relay` in dm-ctl")
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply pkg/terror/error_list.go]: Revised.

File: dm/master/server.go [hunk 9]
```diff
@@ -1886,11 +1888,11 @@ var (
 	retryInterval = time.Second
 )
 
-func (s *Server) waitOperationOk(ctx context.Context, cli workerrpc.Client, name string, opLogID int64) error {
+func (s *Server) waitOperationOk(ctx context.Context, cli workerrpc.Client, taskName, workerID string, opLogID int64) error {
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply dm/master/server.go]: 1. Log added.
2. Added.

File: dm/master/server.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply dm/master/server.go]: Can we use `worker1` in L337?

**Comment by lichunzhu** [comment]
Updated @csuzhangxc

File: dm/master/server.go [hunk 10]
```diff
@@ -324,7 +324,7 @@ func (s *Server) OperateTask(ctx context.Context, req *pb.OperateTaskRequest) (*
 	var wg sync.WaitGroup
 	for _, worker := range workers {
 		// https://golang.org/doc/faq#closures_and_goroutines
-		worker := worker
+		worker1 := worker
```
Comment (csuzhangxc) [inline dm/master/server.go]: 🤔 I think you can use `worker1` in the closures, and no need these two lines.

Comment (lichunzhu) [reply dm/master/server.go]: :joy:  Yes, I omit that before. Revised

**Pull request merged**